[google_maps_flutter] Marker clustering support

jokerttu opened the pull request:
Adds the initial support for the marker clustering for Android, iOS and Web platforms.
Clustering is implemented using native google maps utils libraries for [Android](https://github.com/googlemaps/android-maps-utils), [iOS](https://github.com/googlemaps/google-maps-ios-utils) and [Web](https://github.com/googlemaps/js-markerclusterer).

This PR is created from previous PR: https://github.com/flutter/plugins/pull/6752

Resolves https://github.com/flutter/flutter/issues/26863

**Android**:
![image](https://github.com/flutter/packages/assets/5219613/09b84a8e-f05b-4c71-8808-4043a25201f6)

**iOS**:
![image](https://github.com/flutter/packages/assets/5219613/0859cf12-2e8c-4106-b7a7-cd4922a7dd1e)

**Web**:
![image](https://github.com/flutter/packages/assets/5219613/9269d22a-1908-4c2c-a1ab-70addb06d0f2)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

Hixie commented:
@jokerttu Is this still something you are interested in working on? How can we help you?

cedvdb commented:
@dale-lakes If you are interested, anyone is free to use the fork and keep the PR going so it gets merged faster. I don't think that's against the contribution rules.

jokerttu commented:
> @jokerttu is this still something you're working on?

Yes, I am scheduled to work on this issue next week and will update the PR accordingly in the near future. The recent architectural changes in the `google_maps_flutter_web` implementation have affected the integration tests, which I am addressing. Additionally, some new methods introduced in this PR will likely require updates to the native unit tests.

wangela commented:
@stuartmorgan this is ready for review. We inspected the logs for the failing web platform tests and it seems to be reporting based on Android?

jokerttu commented:
Note: that the original PR was reviewed at plugins repository at https://github.com/flutter/plugins/pull/6752
Implementation parts have been mainly same, but small fixes and changes has been done to fix conflicts.
Web tests are rewritten because of changes on the web implementation (web plugin was endorsed).
iOS Google Maps Utils library version is not [pinned to 4.1.0](https://github.com/flutter/packages/blob/45a866eb565b96f54f4d98b5ce886e4501535ddd/packages/google_maps_flutter/google_maps_flutter_ios/ios/google_maps_flutter_ios.podspec#L25) as newer versions do not support iOS platform versions 11 and 12

stuartmorgan commented:
> @stuartmorgan this is ready for review. We inspected the logs for the failing web platform tests and it seems to be reporting based on Android?

This is because there is a weird circular dependency in the tests where google_maps_flutter_web's example depends on google_maps, which then is depending on a non-overridden version of google_maps_flutter_android so things no longer match. Adding a dependency override for _android in `google_maps_flutter_web/example/pubspec.yaml` should allow the tests to run.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster.dart:
This new file needs corresponding tests. They should be pretty straightforward; see https://github.com/flutter/packages/blob/main/packages/google_maps_flutter/google_maps_flutter_platform_interface/test/types/marker_test.dart for instance.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart:
```diff
@@ -142,6 +142,20 @@ abstract class GoogleMapsFlutterPlatform extends PlatformInterface {
+  }) {
+    throw UnimplementedError(
+        'updateClusterManagers() has not been implemented.');
+  }
```
Please add a simple test that this throws unimplemented.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/platform_interface/google_maps_flutter_platform.dart:
```diff
@@ -357,6 +371,11 @@ abstract class GoogleMapsFlutterPlatform extends PlatformInterface {
+  /// A marker icon managed by [ClusterManager] has been tapped.
+  Stream<ClusterTapEvent> onClusterTap({required int mapId}) {
+    throw UnimplementedError('onClusterTap() has not been implemented.');
+  }
```
Same.

(We occasionally have had issues where just having a test calling new methods highlighted a type export problem, for instance, so this isn't as pointless as it may seem 🙂 )

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart:
This also needs the standard map object test suite.

stuartmorgan reviewed (commented):
Thanks for your patience on this; I've finished reviewing the platform interface, and since the overall structure was already reviewed in the previous PR you are good to split that out into the first prequel PR. I've left some comments that you can either address here before splitting, or just address in the new PR.

I should have the rest of the re-review done later today but I wanted to get this part out first so you could work on that in parallel.

stuartmorgan commented:
@ditman Looks like the original never got your review for web; if you could take a look here when you get a chance that would be great.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GMSMarker+Userdata.m:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  NSString *clusterManagerId = self.userData[1];
+  if (clusterManagerId == (id)[NSNull null]) {
```
The cast here is working around an abuse of the type system on the previous line'; assigning something that may be `NSNull` to an `NSString*` is incorrect. The fix for the compiler message you got is to not do that, rather than to suppress it with casts. If you don't know the type when you extract to a local variable, the correct type for that variable is `id`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapController.m:
```diff
@@ -5,6 +5,8 @@
 #import "FLTGoogleMapJSONConversions.h"
 #import "FLTGoogleMapTileOverlayController.h"
+#import "GMSMarker+Userdata.h"
+@import GoogleMapsUtils;
```
See https://google.github.io/styleguide/objcguide.html#order-of-includes

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapController.m:
```diff
@@ -534,28 +560,32 @@ - (void)mapView:(GMSMapView *)mapView idleAtCameraPosition:(GMSCameraPosition *)
+  if ([marker.userData conformsToProtocol:@protocol(GMUCluster)]) {
+    GMUStaticCluster *cluster = marker.userData;
+    [self.clusterManagersController handleTapCluster:cluster];
+    // When NO is returned the map will focus on the cluster
```
Missing punctuation.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/ClusterManagersController.h:
This file should be called FLTClusterManagersController.h. (Some other files in this plugin are incorrect, but new ones should follow the style guide's naming rules.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -4,25 +4,31 @@
 @property(weak, nonatomic) GMSMapView *mapView;
 @property(assign, nonatomic, readwrite) BOOL consumeTapEvents;
+@property(strong, nonatomic) NSString *clusterManagerId;
+@property(strong, nonatomic) NSString *markerId;
```
These need declaration comments (the existing ones missing comments are wrong too, but we shouldn't perpetuate that).

stuartmorgan replied:
...Identifier

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -4,25 +4,31 @@
 @property(strong, nonatomic) GMSMarker *marker;
 @property(weak, nonatomic) GMSMapView *mapView;
 @property(assign, nonatomic, readwrite) BOOL consumeTapEvents;
+@property(strong, nonatomic) NSString *clusterManagerId;
```
The strings should be `copy`, not `strong`.
https://google.github.io/styleguide/objcguide.html#setters-copy-nsstrings

stuartmorgan replied:
...Identifier

stuartmorgan replied:
...Identifier

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -4,25 +4,31 @@
   if (self) {
-    _marker = [GMSMarker markerWithPosition:position];
+    _marker = marker;
+    _markerId = identifier;
```
And this and below should copy.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.h:
```diff
@@ -4,26 +4,29 @@
-                            identifier:(NSString *)identifier
-                               mapView:(GMSMapView *)mapView;
+- (instancetype)initWithMarker:(GMSMarker *)marker
+                    identifier:(NSString *)identifier
```
There are now two identifiers, so this needs to be more specific.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.h:
```diff
@@ -4,26 +4,29 @@
-                               mapView:(GMSMapView *)mapView;
+- (instancetype)initWithMarker:(GMSMarker *)marker
+                    identifier:(NSString *)identifier
+              clusterManagerId:(NSString *)identifier
```
clusterManagerIdentifier

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -83,15 +89,31 @@ - (void)setRotation:(CLLocationDegrees)rotation {
+  // If marker belongs the cluster manager, visibility need to be controlled with the opacity
+  // as the cluster manager controls when marker is on the map and when not.
+  // Alpha value for marker must always be interpreted before visibility value.
+  if (self.clusterManagerId && self.clusterManagerId != (id)[NSNull null]) {
```
Similar casting comment here: if code is passing `NSNull` to the initializer, which says it takes an `NSString*`, then the calling code is wrong and should be fixed.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -224,6 +246,7 @@ - (UIImage *)scaleImage:(UIImage *)image by:(id)scaleParam {
 @interface FLTMarkersController ()
 
 @property(strong, nonatomic) NSMutableDictionary *markerIdentifierToController;
+@property(weak, nonatomic) FLTClusterManagersController *clusterManagersController;
```
Needs a comment.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -232,6 +255,14 @@ @interface FLTMarkersController ()
+                                          mapView:(GMSMapView *)mapView
+                                        registrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+  _clusterManagersController = clusterManagers;
+  return [self initWithMethodChannel:channel mapView:mapView registrar:registrar];
```
Obj-C initializers should always delegate to the most complete initializer; the way to structure this is to have this new method become the core initializer, and the existing one should call it with a `nil` cluster manager controller.

(Both to follow standard conventions so as not to confuse people, but also because if the initializer you are calling here actually returned a different object from `[super init]` the controller set on the previous line would be lost, so this is incorrect as written.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+                                          clusterManagerId:clusterManagerId
+                                                   mapView:self.mapView];
+  [controller interpretMarkerOptions:markerToAdd registrar:self.registrar];
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
```
Same casting note; line 287 is incorrect if this is possible.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
+    GMUClusterManager *clusterManager =
+        [_clusterManagersController getClusterManagerWithIdentifier:clusterManagerId];
+    if (marker && clusterManager != (id)[NSNull null]) {
```
Same.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
 
+- (void)addMarker:(NSDictionary *)markerToAdd {
+  NSString *identifier = markerToAdd[@"markerId"];
+  NSString *clusterManagerId = markerToAdd[@"clusterManagerId"];
```
`clusterManagerIdentifier`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+
+- (void)changeMarker:(NSDictionary *)markerToChange {
+  NSString *identifier = markerToChange[@"markerId"];
+  NSString *clusterManagerId = markerToChange[@"clusterManagerId"];
```
`clusterManagerIdentifier`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+  if (!controller) {
+    return;
+  }
+  NSString *oldClusterManagerId = [controller clusterManagerId];
```
`previousClusterManagerIdentifier`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+    return;
+  }
+  NSString *clusterManagerId = [controller clusterManagerId];
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
```
Same casting issue. If `clusterManagerId` is returning `NSNull` it is doing so incorrectly.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m:
```diff
@@ -247,37 +278,75 @@ - (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+  if (clusterManagerId && clusterManagerId != (id)[NSNull null]) {
+    GMUClusterManager *clusterManager =
+        [_clusterManagersController getClusterManagerWithIdentifier:clusterManagerId];
+    if (controller.marker && clusterManager != (id)[NSNull null]) {
```
(As a general note: it is *very* rare to have a legitimate reason to cast to `(id)` in Obj-C code; any time you find yourself doing it, it's very likely something deeper is wrong.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/google_maps_flutter_ios.podspec:
```diff
@@ -3,7 +3,7 @@
 Pod::Spec.new do |s|
   s.name             = 'google_maps_flutter_ios'
-  s.version          = '0.0.1'
+  s.version          = '0.0.2'
```
We never change this.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/google_maps_flutter_ios.podspec:
```diff
@@ -22,8 +22,14 @@ Downloaded by pub (not CocoaPods).
   # has been confirmed to be compatible via an example in examples/. See discussion
   # in https://github.com/flutter/flutter/issues/86820 for why this is so broad.
   s.dependency 'GoogleMaps', '< 8.0'
+  s.dependency 'Google-Maps-iOS-Utils', '4.1.0'
```
It's probably worth adding a comment here saying that this should be updated once the plugin requires 13.0+.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart:
```diff
@@ -587,6 +622,32 @@ class GoogleMapsFlutterIOS extends GoogleMapsFlutterPlatform {
+
+  /// Parses cluster data from dynamic json objects and returns [Cluster] object.
+  /// Used by `cluster#onTap` method call handler and inspectors [getClusters] response parser.
+  static Cluster parseCluster(
```
Shouldn't this be private?

illuminati1911 replied:
`GoogleMapsInspectorIOS`class is using it and needs it to be public.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/lib/src/utils/cluster_manager.dart:
These utilities should have tests.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/example/web/index.html:
```diff
@@ -7,6 +7,7 @@
         <title>Browser Tests</title>
         <!-- This API key comes from: go/flutter-maps-web-tests-api-key (GCP project: flutter-infra) -->
         <script src="https://maps.googleapis.com/maps/api/js?key=AIzaSyAa9cRBkhuxGq3Xw3HPz8SPwaVOhRmm7kk&libraries=geometry"></script>
+        <script src="https://unpkg.com/@googlemaps/markerclusterer@2.5.0/dist/index.min.js"></script>
```
(If there's any change based on my comment on the app-facing package, we should make sure to make it here as well.)

jokerttu replied:
This is now updated to match the cdn used in web implementation as well.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart:
```diff
@@ -0,0 +1,200 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: public_member_api_docs, non_constant_identifier_names
```
Why are we bulk-ignoring analysis in non-test code?

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart:
This file should likely have Dart unit tests.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/example/web/index.html:
```diff
@@ -38,6 +38,7 @@
 
   <!-- This API key comes from: go/flutter-maps-web-tests-api-key (GCP project: flutter-infra) -->
   <script src="https://maps.googleapis.com/maps/api/js?key=AIzaSyAa9cRBkhuxGq3Xw3HPz8SPwaVOhRmm7kk&libraries=geometry"></script>
+  <script src="https://unpkg.com/@googlemaps/markerclusterer@2.5.0/dist/index.min.js"></script>
```
@ditman is probably the best person to answer that question.

ditman replied:
The script tag seems to be a recommended way[^1] of injecting the library, so this will have to do. unpkg.com should be familiar to some flutter web users, because for a while that was our way of using canvaskit!

As @stuartmorgan said it's not ideal because of the new network dependency, but I'd rather have this than dealing with the shrink-wrapping of the library as an asset of the package.

[^1]: https://github.com/googlemaps/js-markerclusterer?tab=readme-ov-file#install

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:
```diff
@@ -0,0 +1,222 @@
+  public void removeItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
```
ClusterManagerUpdates are handled before MarkerUpdates and there might be situation where clusterManager is removed and markers are still updating as these updates are unawaited method calls from the dart side.
https://github.com/flutter/packages/pull/4319/files#diff-9bd2e43e18b051bf0a22c1fc3430348ca4b37755cdfd9b4dbb75a98ba156786bR373

I can test if it is possible that there is conditions where previous update is still updating huge set of markerupdates, when next update comes that removes the clustermanager.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:
```diff
@@ -0,0 +1,222 @@
+  public void addItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
```
Same issue: https://github.com/flutter/packages/pull/4319/files#r1432618672

apackin commented:
Any progress updates on when we can expect this to come out?

hellohuanlin commented:
@jokerttu is this ready for another review?

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/google_maps_flutter_web.dart:
```diff
@@ -31,6 +32,7 @@ part 'src/convert.dart';
 part 'src/google_maps_controller.dart';
 part 'src/google_maps_flutter_web.dart';
 part 'src/marker.dart';
+part 'src/marker_clustering.dart';
```
Does this really really need to be a `part`? Can it be `import`ed?

jokerttu replied:
Most likely it can be imported as well with minor changes, using part was chosen to follow the design of this package as everything else is implemented as partof syntax.

jokerttu replied:
marker_clustering.dart is now imported.
it uses shared conversion methods from the convert.dart file.
these methods were private, and I needed to change them as public to be able to use them from the maker_clustering.dart file.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart:
```diff
@@ -0,0 +1,205 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: public_member_api_docs, non_constant_identifier_names
```
```suggestion
// ignore_for_file: non_constant_identifier_names
```

I guess `non_constant_identifier_names` is needed to keep the capitalization of things coming from the JS library?

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart:
```diff
@@ -0,0 +1,205 @@
+  }
+}
+
+@JS()
+external ClusterClickHandler defaultOnClusterClickHandler;
```
I have a suspicion that these JS-interop bits need to be split to its own specific package (similar to `google_maps` or `google_identity_services_web`), actually the  new package should only have dependencies in `google_maps`, and then be used from `google_maps_flutter_web`, which would own converting between the JS-types and the ones from the `platform_interface`)

jokerttu replied:
Marker clustering is quite tightly mingled with the marker controller, so separating the whole markerclustering to separate package might be a bit tricky.
Are you suggesting that only the js interops are separated to separata package, or whole clustering support, as latter might be quite tricky to do.

ditman replied:
@jokerttu I meant only the `@JS` bits. It'd be a tiny package, but not baked into this package. If you split it to a separate file in this PR, we can move it to a separate package later.

ditman commented on packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart:
```diff
@@ -0,0 +1,205 @@
+@JS()
+external ClusterClickHandler defaultOnClusterClickHandler;
+
+@JS()
+@anonymous
+class MarkerClustererOptions {
```
This needs some of the new `@staticInterop`, coming from `dart:js_interop`, instead of `package:js` (there's a migration to the new interop [incoming](https://github.com/flutter/packages/pull/5254/))

jokerttu replied:
I'll check what changes this would need

jokerttu replied:
I couldn't find a way to use @staticInterop with objects from the google_maps library because the package does not utilize static interop. Additionally, I did not find a clear method to bypass this issue, as clustering is tightly coupled with the Google Maps javascript objects. If the objects within the google_maps package were implemented using @staticInterop, it should be possible to use these objects as function parameters directly.

jokerttu replied:
@ditman as you mentioned the [incoming](https://github.com/flutter/packages/pull/5254/) update that should also update google_maps to version 7.x.x having `package:web `support, might be that moving to use `@staticInterop` is possible after that.

ditman replied:
Yeah, in fact, I've been told that I need to update the JS-interop on maps (and others) to package:web 0.5.0, which doesn't use `@staticInterop` anymore, and instead extension types, like this: https://github.com/flutter/flutter/pull/143274 (so consider my original comment obsolete, I'll work this week on updating the migration PR!)

jokerttu replied:
@ditman given the recent updates and the fact that this PR touches on 5 different plugins, leading often to merge conflicts, (especially now that the changelogs are up to date), I wanted to inquire if there's anything more needed from my side to expedite the finalization of this PR for web?

ditman reviewed (commented):
Would it be possible to split the js-interop API layer of `js-markerclusterer` to its own separate package, that only has dependencies on a14n's `google_maps`?

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:
```diff
@@ -807,6 +856,19 @@ private void updateInitialMarkers() {
+  public void setInitialClusterManagers(Object initialClusterManagers) {
+    ArrayList<?> clusterManagers = (ArrayList<?>) initialClusterManagers;
+    this.initialClusterManagers = clusterManagers != null ? new ArrayList<>(clusterManagers) : null;
+    if (googleMap != null) {
```
Same logic id used for other updates as well (eg. in setInitialPolygons method).
in this method `this.initialClusterManagers` value is set, and it is used when googleMap object is initialized.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -221,6 +223,41 @@ static Object latLngToJson(LatLng latLng) {
     return Arrays.asList(latLng.latitude, latLng.longitude);
   }
 
+  static Object clustersToJson(
```
Done

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:
```diff
@@ -221,6 +223,41 @@ static Object latLngToJson(LatLng latLng) {
+    return data;
+  }
+
+  static Object clusterToJson(String clusterManagerId, Cluster<MarkerBuilder> cluster) {
```
Bundled with clustersToJson as it calls clusterToJson internally

jokerttu commented:
@stuartmorgan @ditman 
I've addressed most of the review comments and made the corresponding updates to the PR (with help of @illuminati1911). However, a few discussions remain open, and I'm actively seeking further input on those.

Regarding the web implementation, I encountered challenges with using @staticInterop annotations for methods that involve google-maps interop objects as parameters. Since google-maps objects do not utilize @staticInterop, it appears we cannot proceed in this direction for the time being. I'm open to suggestions or alternative approaches to resolve this issue.

stuartmorgan commented:
> I've finished reviewing the platform interface, and since the overall structure was already reviewed in the previous PR you are good to split that out into the first prequel PR. I've left some comments that you can either address here before splitting, or just address in the new PR.

It looks like this still hasn't been split out?

jokerttu commented:
@stuartmorgan 
> It looks like this still hasn't been split out?

The review comments were resolved in this PR before splitting so that the platform-specific implementations would not stretch too far in time after the platform interface update.

Here is the first prequel PR for `google_maps_flutter_platform_interface`: https://github.com/flutter/packages/pull/6158

stuartmorgan commented:
> However, a few discussions remain open, and I'm actively seeking further input on those.

Was there anything other than web-specific questions? I looked back through the comments and didn't find any unresolved discussions with questions other than the web ones, but I may have missed something.

jokerttu commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GMSMarker+Userdata.h:
Implementation has changed how userdata is handled.

jokerttu commented:
> > However, a few discussions remain open, and I'm actively seeking further input on those.
> 
> Was there anything other than web-specific questions? I looked back through the comments and didn't find any unresolved discussions with questions other than the web ones, but I may have missed something.

Yes, checked and only web-specific discussions are still open.

jokerttu commented:
@stuartmorgan added prequel PR:s for platforms as well as **draft**, as the platform_interface (https://github.com/flutter/packages/pull/6158) PR has not yet landed:
Android: https://github.com/flutter/packages/pull/6185
iOS: https://github.com/flutter/packages/pull/6186
Web: https://github.com/flutter/packages/pull/6187 (@ditman)

jokerttu commented:
Converted to draft to avoid unnecessary reviews before the platform implementations are landed.

ParkBud commented:
@jokerttu any updates on this 🥺 Really looking forward to it. Really appreciate you for taking this on!

jokerttu commented:
> @jokerttu any updates on this 🥺 Really looking forward to it. Really appreciate you for taking this on!

The work is ongoing on platform implementation PR:s
https://github.com/flutter/packages/pull/6185
https://github.com/flutter/packages/pull/6186
https://github.com/flutter/packages/pull/6187

stuartmorgan commented:
Status update from triage: Blocked on landing the iOS platform implementation, which is turn blocked on Google Maps SDK-level issues.

Mohit8G commented:
Any update?

jokerttu commented:
@stuartmorgan this repo now holds only app facing package changes, and should be ready for final reviews.

stuartmorgan reviewed (approved):
LGTM!

Mohit8G commented:
https://medium.com/@rishyash8/clustering-support-in-google-map-flutter-2024-d8b1c68e515b

Basic implementation as example code is not updated in google_map_flutter on pub.dev

Rishyash commented:
https://medium.com/@rishyash8/clustering-support-in-google-map-flutter-2024-d8b1c68e515b

Basic implementation as example code is not updated in google_map_flutter on pub.dev

stuartmorgan commented:
Pub.dev only shows a single file in the example tab, and the example app for this package is quite large; all the individual features are on sub-pages, and clustering is no different. The example app can be viewed or run locally to see specific features.

Mohit8G commented:
Hey Everyone, just gone through whole code for clustering part for flutter, cant find Custom Cluster Renderer for customising the cluster icon. Any update on that, like are we picking it in near future.

stuartmorgan commented:
@Mohit8G This is a PR that has landed, so it cannot be changed. If you have follow-up feature requests, please file them in the issue tracker.

stuartmorgan commented:
> yep, but the example which was present in the package was quite complex to understand as a beginner so created a simple example for that.

"I wanted to make a simpler example for beginners" and "the updated example code was not published to pub.dev" are very different statements; the latter would have indicated some kind of serious bug in the release process. Please don't post to a PR to promote your post but describe it as if it were a bug in the PR or release.

stuartmorgan commented:
Temporarily locking, as this is attracting a lot of off-topic comments. Discussion on a PR should be about the code in the PR, not things like follow-up requests.
